### ISSUE TITLE
Return information about the applied update from `Y.applyUpdate`

### ISSUE THREAD

**Issue by ankon** [yjs/yjs#405]
**Checklist**

[x] Are you reporting a bug? Use github issues for bug reports and feature requests. For general questions, please use https://discuss.yjs.dev/
[x] Try to report your issue in the correct repository. Yjs consists of many modules. When in doubt, report it to https://github.com/yjs/yjs/issues/

**Is your feature request related to a problem? Please describe.**

We recently found various cases of "corruption", caused by missing updates. As part of improving the system we realized that it would be good if we could detect immediately when applying an update whether that update has any missing dependencies. It seems that part of applying an update knows that, https://github.com/yjs/yjs/blob/2ec19defcb244fa32d0033f050c38c0f57a01e59/src/utils/encoding.js#L223 `integrateStructs` can return the "missing" information. But, unfortunately the `applyUpdate` function doesn't expose that.

**Describe the solution you'd like**

Return information about whether the `applyUpdate` call "worked", and potentially what is still missing.

**Describe alternatives you've considered**

Not considered yet, but I suspect we could somehow look at the metadata of an update, and somehow check the metadata of the current document, and work out whether the update is "the next one" or whether there is anything else "in between" that would still be needed.

**Comment by dmonad** [comment]
Implementing something like this would be too complex. Furthermore, it is completely fine to apply an update that depends on a missing update. Once the missing update is applied, the first update will be applied implicitly.

If part of an update cannot be applied, you can find it on `ydoc.store.missing*`. You can copy these properties and set them to null if that is something you want to do (not recommended; only for debugging). These properties also contain some information on which updates are missing. In the future, I might want to rename/move/optimize these properties. So consider them non-stable.

To be clear: It is definitely not feasible to look at an update and check which updates (or lamport timestamps) are missing. Therefore, I'll close this issue. But feel free to continue asking questions.

**Comment by ankon** [comment]
Thanks for your reply!

> [...] it is completely fine to apply an update that depends on a missing update. Once the missing update is applied, the first update will be applied implicitly.

Indeed. Our problem is that we actually found ways caused by network infrastructure/issues to actually lose these updates completely, i.e. these missing updates will never get applied. Our goal with this request was also not to provide some magic way of fixing it, but rather to have a way to detect such a problem in tests, and to (for instance) log an error into the browser console or help a support tool to do a "integrity check" (walk over the complete set of known issues, report whether there appear to be any missing ones).

> If part of an update cannot be applied, you can find it on `ydoc.store.missing*`. You can copy these properties and set them to null if that is something you want to do (not recommended; only for debugging). These properties also contain some information on which updates are missing.

This actually sounds a lot like what we need, and after some thinking I also realized that the set of "missing" things isn't just a return value from the `apply` operation, but rather a property on the document itself. I'll have a look at what information is available there, but I'm quite optimistic that I'll be able to find what I am looking for. Thanks for the pointer!

> So consider them non-stable.

Of course!
